README.md
[![Release](https://img.shields.io/github/v/release/scaleflex/react-cloudimage-responsive)](https://github.com/scaleflex/js-cloudimage-responsive/releases)
[![Free plan](https://img.shields.io/badge/price-includes%20free%20plan-green.svg)](https://www.cloudimage.io/en/home#b38181a6-b9c8-4015-9742-7b1a1ad382d5)
[![Contributions welcome](https://img.shields.io/badge/contributions-welcome-orange.svg)](#contributing)
[![License](https://img.shields.io/badge/license-MIT-blue.svg)](https://opensource.org/licenses/MIT)
[![Scaleflex team](https://img.shields.io/badge/%3C%2F%3E%20with%20%E2%99%A5%20by-the%20Scaleflex%20team-6986fa.svg)](https://www.scaleflex.com/en/home)
[![Cloudimage](https://img.shields.io/badge/Powered%20by-cloudimage-blue)](https://www.cloudimage.io/en/home)

## VERSIONS

* __Low Quality Preview__
* [__Blur Hash__](https://github.com/scaleflex/react-cloudimage-responsive-blur-hash)
* [__Plain (CSS free)__](https://github.com/scaleflex/react-cloudimage-responsive-plain)

<p align="center">
	<a href="https://www.cloudimage.io/#gh-light-mode-only">
		<img
			alt="The Lounge"
			src="https://scaleflex.cloudimg.io/v7/cloudimage.io/LOGO+WITH+SCALEFLEX-01.png?vh=f6080d&w=350">
	</a>
		<a href="https://www.cloudimage.io/#gh-dark-mode-only">
		<img
			alt="The Lounge"
			src="https://scaleflex.cloudimg.io/v7/cloudimage.io/cloudimage-logo-light.png?vh=b798ab&w=350">
	</a>
</p>

<h1 align="center">
   React Cloudimage Responsive | Cloudimage v7
</h1>

<h3 align="center">
(low quality preview)
</h3>

<p align="center">
	<strong>
		<a href="#table_of_contents">Docs</a>
		•
		<a href="https://scaleflex.github.io/react-cloudimage-responsive/" target="_blank">Demo</a>
		•
		<a href="https://codesandbox.io/s/react-cloudimage-responsive-example-re0y6m" target="_blank">Code Sandbox</a>
		•
		<a href="https://medium.com/@dmitry_82269/responsive-images-in-2019-now-easier-than-ever-b76e5a43c074" target="_blank">Why?</a>
	</strong>
</p>

This plugin detects the width of any image container as well as the device pixel ratio
density to load the optimal image size needed.
Images are resized on-the-fly via the <a href="https://cloudimage.io" target="_blank">Cloudimage service</a>, thus offering a comprehensive
automated image optimization service.

When an image is first loaded on your website or mobile app,
Cloudimage's resizing servers will download the origin image from
the source, resize it for the client's screen size and deliver to your users through one or multiple
Content Delivery Networks (CDNs). The generated image formats are cached in the CDN and will be delivered rocket fast on any subsequent request.

**NOTE:** Your original (master) images have to be stored on a server
or storage bucket (S3, Google Cloud, Azure Blob...) reachable over
HTTP or HTTPS by Cloudimage. If you want to upload your master images to
Cloudimage, contact us at
[hello@cloudimage.io](mailto:hello@cloudimage.io).

<p align="center">
	<img
		alt="The Lounge"
		src="https://cdn.scaleflex.it/filerobot/cloudimage-process.jpg">
</p>

## Table of contents

* [Demo](#demo)
* [Responsive plugins family](#plugin_family)
* [Compatibility](#compatibility)
* [Requirements](#requirements)
* [Step 1: Installation](#installation)
* [Step 2: Initialize](#initialize)
* [Step 3: Implement](#implement)
* [Configuration](#configuration)
* [Image properties](#image_properties)
* [Browser support](#browser_support)
* [Filerobot UI Family](#ui_family)
* [Contributing](#contributing)
* [License](#license)


## <a name="demo"></a> Demo

To see the Cloudimage Responsive plugin in action, please check out the
[Demo page](https://scaleflex.github.io/react-cloudimage-responsive/).
Play with your browser's window size and observe your
Inspector's Network tab to see how Cloudimage delivers the optimal
image size to your browser, hence accelerating the overall page
loading time.

## <a name="plugin_family"></a> Responsive plugins family
In order to use Cloudimage responsive plugins on your single-page application, please check out Cloudimage responsive plugins for vanilla Javascript and the most popular Javascript frameworks.

<details>
<summary>React</summary>
<a href="https://github.com/scaleflex/react-cloudimage-responsive">React Cloudimage Responsive (Low Quality Preview)</a><br/>
<a href="https://github.com/scaleflex/react-cloudimage-responsive-blur-hash">React Cloudimage Responsive (Blur-hash)</a><br/>
<a href="https://github.com/scaleflex/react-cloudimage-responsive-plain">React Cloudimage Responsive (Plain)</a>
</details>

<details>
<summary>Next.js</summary>
<a href="https://github.com/scaleflex/next-cloudimage-responsive">Vue Cloudimage Responsive (Low Quality Preview)</a><br/>
</details>

<details>
<summary>Vue.js</summary>
<a href="https://github.com/scaleflex/vue-cloudimage-responsive">Vue Cloudimage Responsive (Low Quality Preview)</a><br/>
<a href="https://github.com/scaleflex/vue-cloudimage-responsive-blur-hash">Vue Cloudimage Responsive (Blur-hash)</a><br/>
<a href="https://github.com/scaleflex/vue-cloudimage-responsive-plain">Vue Cloudimage Responsive (Plain)</a>
</details>

<details>
<summary>Angular</summary>
<a href="https://github.com/scaleflex/ng-cloudimage-responsive">Angular Cloudimage Responsive (Low Quality Preview)</a><br/>
</details>

## <a name="compatibility"></a> Compatibility
| Version | React Version |
|---------|--------------|
| >3.6.3   | >=18.0.0 |
| <=3.6.3   | <=17.x.x |

## <a name="requirements"/> Requirements

To use the Cloudimage Responsive plugin, you will need a
Cloudimage token to deliver your images over CDN. Don't worry, it only takes seconds to get one by
registering [here](https://www.cloudimage.io/en/register_page).
Once your token is created, you can configure it as described below.
This token allows you to use 25GB of image cache and 25GB of worldwide
CDN traffic per month for free.

## <a name="installation"></a>Step 1: Installation

using npm

```
$ npm install --save react-cloudimage-responsive
```

## <a name="initialize"></a>Step 2: Initialize

After installing the react-cloudimage-responsive lib, simply initialize it with your **token** and the **baseURL**
of your image storage with **CloudimageProvider**:

```jsx
import React from 'react';
import { render } from 'react-dom';
import Img, { CloudimageProvider } from 'react-cloudimage-responsive';

const cloudimageConfig = {
  token: 'demo',
  baseURL: 'https://cdn.scaleflex.it/demo/'
};

const App = () => {
  return (
    <CloudimageProvider config={cloudimageConfig}>
      <h1>Simple demo of react-cloudimage-responsive</h1>
      <Img src="img.jpg" alt="Demo image" ratio={1.5}/>
    </CloudimageProvider>
  );
};

render(<App />, document.body);
```

## <a name="implement"></a>Step 3: Implement it

### Img component:

```html
<Img src="img.jpg" alt="Demo image" ratio={1.5}/>
```

NOTE: "ratio" is recommended to prevent page layout jumping. The parameter is used to calculate image height to hold
the image position while image is loading.

<a href="https://codesandbox.io/s/react-cloudimage-responsive-example-9e9rb5"><img src="https://codesandbox.io/static/img/play-codesandbox.svg" alt="edit in codesandbox"/></a>

### BackgroundImg component:

```html
<BackgroundImg src="img.jpg">
 {'Your content...'}
</BackgroundImg>
```

<a href="https://codesandbox.io/s/react-cloudimage-responsive-example-fseygn"><img src="https://codesandbox.io/static/img/play-codesandbox.svg" alt="edeit in codesandbox"/></a>

## <a name="configuration"></a> Config

### token

###### Type: **String** | Default: **"demo"** | _required_

Your Cloudimage customer token.
[Subscribe](https://www.cloudimage.io/en/register_page) for a
Cloudimage account to get one. The subscription takes less than a
minute and is totally free.

### customDomain

###### Type: **String** | Default: **"cloudimage.io"** | _optional_

If you use a custom CNAME for your cloudimage integration, you can set it here.

Note: this will disregard your token above as this should be built into the CNAME entry.

### baseURL

###### Type: **String** | Default: **"/"** | _optional_

Your image folder on server, this alows to shorten your origin image URLs.

### apiVersion

###### Type: **String** |Default: **'v7'** | _optional_

Allow to use a specific version of API.

- set a specific version of API
```javascript
const cloudimageConfig = {
  token: 'demo',
  baseURL: 'https://cdn.scaleflex.it/demo/',
  apiVersion: 'v7'                            // optional
};
```
- disable API version
```javascript
const cloudimageConfig = {
  token: 'demo',
  baseURL: 'https://cdn.scaleflex.it/demo/',
  apiVersion: null                            // optional
};
```


### doNotReplaceURL

###### Type: **bool** | Default: **false**

If set to **true** the plugin will only add query params to the given source of image.

### <a name="lazy_loading_config"></a>lazyLoading

###### Type: **Bool** | Default: **true** | _optional_

Only images close to the client's viewport will be loaded, hence accelerating the page loading time. The plugin uses
[react-lazyload](https://github.com/twobin/react-lazyload) library to achieve it.

### lazyLoadOffset

######  Type: **Number/Array(Number)** | Default: **100**

Say if you want to preload a component even if it's 100px below the viewport (user have to scroll 100px more to see this component), you can set offset props to 100. On the other hand, if you want to delay loading a component even if it's top edge has already appeared at viewport, set offset to negative number.

Library supports horizontal lazy load out of the box. So when you provide this prop with number like 100 it will automatically set left edge offset to 100 and top edge to 100;

If you provide this prop with array like [100, 200], it will set left edge offset to 100 and top offset to 200.

### params

###### Type: **String** | Default: **'org_if_sml=1'** | _optional_

Applies default Cloudimage operations/ filters to your image, e.g. brightness, contrast, rotation...
Multiple params can be applied, separated by "```&```" e.g. wat_scale=35&wat_gravity=northeast&wat_pad=10&grey=1

```javascript
params: 'org_if_sml=1'
```

#### alternative syntax: type: **Object**

```javascript
params: {
    org_if_sml: 1,
    grey: 1,
    ...
}
```

[Full cloudimage v7 documentation here.](https://docs.cloudimage.io/go/cloudimage-documentation-v7/en/introduction)


### placeholderBackground

###### Type: **String** | Default: **'#f4f4f4'** | _optional_

Placeholder coloured background while the image is loading or use it to set your custom placeholder image or gif

for example

```placeholderBackground: "url('https:/https://cdn.scaleflex.it/filerobot/red-loader.gif') 50% 50% no-repeat"```

### lowQualityPreview

###### Type: **Object**

* `lowQualityPreview.minImgWidth` number (default: 400) - minimum width of an image to load a low-quality preview image

Example:

```javascript
lowQualityPreview: {
  minImgWidth = 400
}
```

### presets

###### Type: **Object**

Default:

```javascript
const cloudimageConfig = {
  token: 'demo',
  baseUrl: 'https://cdn.scaleflex.it/demo/',
  ...
  presets: {
      xs: '(max-width: 575px)', // up to 575    PHONE
      sm: '(min-width: 576px)', // 576 - 767    PHABLET
      md: '(min-width: 768px)', // 768 - 991    TABLET
      lg: '(min-width: 992px)', // 992 - 1199   SMALL_LAPTOP_SCREEN
      xl: '(min-width: 1200px)' // from 1200    USUALSCREEN
  }
};
```

Breakpoints shortcuts to use in image size property, can be overridden.

### limitFactor

###### Type: **Number** | Default: **100** | _optional_

Rounds up the size of an image to the nearest limitFactor value.

For example:
* for an image with width **358px** and limitFactor equal to **100**, the plugin will round up to 400px;
* for an image with width **358px** and limitFactor equal to **5**, the plugin will round up to 360px.

### devicePixelRatioList

###### Type: **[Number,...]** | Default: **[1, 1.5, 2]** | _optional_

List of supported device pixel ratios. If there is no need to support retina devices, you should set an empty array `devicePixelRatioList: []`.

### delay

###### Type: **Number** | _optional_

Delay for processing an image after rendering component.

**NOTE:** normally, the parameter is not needed but in some cases with integrating third-party libraries, it can fix wrong
calculation of image container.

### ImageSizeAttributes

###### Type: **String** | possible values: 'use', 'ignore', 'take-ratio' | Default: **'use'** 

If width and height attributes are set:

**use** - width & height attributes values will be used to calculate image size (according to user's DPR) and **ratio**. 

**take-ratio** - width & height attributes values will be used only to calculate **ratio**.

**ignore** - width & height attributes will be ignored.

If width and height attributes are NOT set, image container size will be detected to calculate result image size (according to user's DPR)

*Note*: If only width or height attributes is set, ratio is going to be taken from ci-ratio image attribute

## <a name="image_properties"></a> Image properties

### src

###### Type: **String** | Default: **undefined** | _required_

Original image hosted on your web server. You can use absolute path or
relative to the baseUrl in your config.

**NOTES:**

The plugin uses a special algorithm to detect the width of image container and set the image size accordingly.
This is the recommended way of using the Cloudimage Responsive plugin.

### onImgLoad

###### Type: **function** | Default: **undefined**

A function to be executed after the image is loaded

```jsx
<Img
  src="dino-reichmuth-1.jpg"
  onImgLoad={(params)=>{console.log(params);}}/>
```
### doNotReplaceURL

###### Type: **Boolean** | Default: **false** | _optional_

If set to true, the plugin will only add query parameters to the provided image source URL.

<a href="https://codesandbox.io/s/react-cloudimage-responsive-example-6xl8pw"><img src="https://codesandbox.io/static/img/play-codesandbox.svg" alt="edeit in codesandbox"/></a>

### disableAnimation

###### Type: **Boolean** | Default: **false** | _optional_

If set, the plugin will disable the animation for the Img and render only a single img tag.

### width

###### Type: **String** (e.g. 300px, 20vw) | Default: **undefined**

If set, the plugin will use width as fixed value and change only according to the device pixel ratio.

### height

###### Type: **String** (e.g. 300px, 20vh) | Default: **undefined**

If it's set, the plugin will use height as fixed value and change only according to the device pixel ratio.

### params

###### Type: **String** | Default: **undefined** | _optional_

You can apply any Cloudimage operations/filters to your image, like brightness, contrast, rotation, etc.
Multiple parameters can be applied, separated by "```&```" e.g. **wat_scale=35&wat_gravity=northeast&wat_pad=10&grey=1**

```javascript
params="gray=1&bright=10"
```

#### alternative syntax: type: **Object**

```javascript
params={{
    bright: 10,
    grey: 1,
    ...
}}
```

[Full cloudimage v7 documentation here.](https://docs.cloudimage.io/go/cloudimage-documentation-v7/en/introduction)

### sizes

###### Type: **Object** | Default: **undefined**

**{preset breakpoint (xs,sm, md,lg,xl) or media query + ' ' + image params}**:

```jsx
<Img
  src="dino-reichmuth-1.jpg"
  sizes={{
    '(max-width: 575px)': { w: 400, h: 150 },
    '(min-width): 576px)': { r: 1 },
    '(min-width: 620px)': { h: 560 },
    '(min-width: 768px)': { w: '50vw' },
    '(min-width: 992px)': { w: '55vw', h: 300 },
    '(min-width: 1200px)': { w: 1200 }
 }}/>
```

You can drop some breakpoints, for example:

```jsx
<Img
  src="dino-reichmuth-1.jpg"
  sizes={{
      sm: { w: 400, h: 200 },
      '(min-width: 620px)': { w: 200, h: 60 }
 }}/>
```

##### new experimental syntax

md: { w: '40vw', h: 350 } or md: { w: 250, h: '20vh' }

adds possibility to use fixed height or width and change dynamically other dimension

**NOTE:** if size is not set, the plugin uses a special algorithm to
detect the width of image container and set the image size accordingly. This is the recommended way of using the Cloudimage Responsive plugin.

### ratio

###### Type: **Number** | _optional_

It is recommended to set this parameter to prevent page layout jumping. It is used to calculate the image height to hold
the position of the image while it is loading.

### lazyLoading

###### Type: **Bool** | _optional_

It makes possible to disable lazyLoading on a per-image basis.

### lazyLoadConfig

###### Type: **Object** | _optional_

The lazyLoad configuration to [LazyLoad](https://github.com/twobin/react-lazyload#props) component. 

To see the full Cloudimage documentation, [click here](https://docs.cloudimage.io/go/cloudimage-documentation).

## <a name="browser_support"></a>Browser support

Tested in all modern browsers and IE 11.

## <a name="ui_family"></a>Filerobot UI Familiy

* [JS Cloudimage Responsive](https://github.com/scaleflex/js-cloudimage-responsive)
* [Angular Cloudimage Responsive](https://github.com/scaleflex/ng-cloudimage-responsive)
* [JS Cloudimage 360 view](https://github.com/scaleflex/js-cloudimage-360-view)
* [Image Editor](https://github.com/scaleflex/filerobot-image-editor)
* [Uploader](https://github.com/scaleflex/filerobot-uploader)

## <a name="contributing"></a>Contributing!

All contributions are super welcome!


## <a name="license"></a>License
React Cloudimage Responsive is provided under the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Types of changes:
- `Added` for new features.
- `Changed` for changes in existing functionality.
- `Deprecated` for soon-to-be removed features.
- `Removed` for now removed features.
- `Fixed` for any bug fixes.
- `Security` in case of vulnerabilities.

> Date format: YYYY-MM-DD

> If we have some "Breaking changes" we can mark it in message by `**BREAKING**` preffix, like:
> `- **BREAKING**: Some message`

-------------

## TODOs
> Todo list for future

- ...

-------------

-------------
## 3.6.6 - 2023-10-18
### Fixed
- Replace `react-lazyload` plugin with `react-lazy-load` to support React 18.

## 3.6.5 - 2023-04-18
### Fixed
- Remove CVE vulnerabilities

## 3.6.4 - 2022-08-04
### Fixed
- prop-types dependency

## 3.6.3 - 2022-08-04
### Changed
- Support React 18
### Fixed
- Removed deprecated methods

## 3.6.2 - 2022-04-29
### Fixed
- Low preview image width

## 3.6.1 - 2022-04-26
### Fixed
- Low preview image width

## 3.6.0 - 2022-04-25
### Changed
- Update documentations

## 3.6.0-beta.20 - 2022-04-18
### Added
- possibility to render unwrapped image using `disableAnimation` prop

## 3.6.0-beta.19 - 2022-04-18
### Added
- possibility to add `doNotReplaceURL` to single image

## 3.6.0-beta.17 - 2022-02-09
### Fixed
- change devicePixelRatioList to [1, 1.5, 2]

## 3.6.0-beta.16 - 2022-01-18
### Fixed
- prevent the lazyLoading property from being spread onto the img tag

## 3.6.0-beta.15 - 2021-01-04
### Fixed
- removed hardcoded ci_info query parameter to fix image caching

## 3.6.0-beta.14 - 2021-12-22
### Fixed
-  img-src starts with "//"

## 3.6.0-beta.13 - 2021-11-09
### Fixed
- possibility to remove API versions

## 3.6.0-beta.11 - 2021-10-14
### Changed
- better support for SSR (render placeholders with correct sizes via SSR stage)


## 3.5.1 - 2021-07-27
### Fixed
- domain property in config


## 3.5.0 - 2021-07-23
### Added
- new property: **customDomain**


## 3.4.0 - 2021-06-19
### Deprecated
Property **ignoreNodeImgSize** is deprecated. Use **imageSizeAttributes: 'ignore'** instead
### Added
- new property: **imageSizeAttributes**


## 3.3.2 - 2020-06-26
### Added
- onImgLoad handler to Img and BackgroundImg components


## 3.3.2 - 2020-09-09
### Added
- allow unmounting the provider


## 3.3.1 - 2020-06-10
### Added
- delay property to handle cases when we need wait till some layout is loaded first


## 3.3.0 - 2020-04-29
### Added
- missing properties in the documentation (README), fix some mistakes
### Fixed
- problem with not setting ratio on preview load when no image is set


## 3.2.4 - 2020-04-28

### Fixed
- problem with srcset attribute

### Added
- add SSR support

## 3.2.3 - 2020-04-28

### Changed
- improve logic for bound resize

## 3.2.2 - 2020-04-25

### Added
- add possibility to configure minLowQualityWidth

### Changed
- improve logic for bound resize

## 3.2.1 - 2020-04-25

### Fixed
- problem with not setting ratio
- if use params as objects remove from html

## 3.1.0 - 2020-04-02

### Changed
- babel version from 6.x.x to 7.x.x

## 1.2.0 - 2020-04-22

### Changed
- improve logic structure / core refactoring

## 1.1.1 - 2020-04-16

### Added
- separate build without polyfills to support Gatsby


src/provider.js
import {
  createContext, useEffect, useState, useMemo,
} from 'react';
import { debounce } from 'throttle-debounce';
import { CONSTANTS, processParams } from 'cloudimage-responsive-utils';


const CloudimageContext = createContext({ cloudImageConfig: {} });

function CloudimageProvider({ config = {}, children } = {}) {
  const {
    token = '',
    domain = config.customDomain || 'cloudimg.io',
    customDomain = false,
    lazyLoading = false,
    lazyLoadOffset = 100,
    placeholderBackground = '#f4f4f4',
    baseUrl, // to support old name
    baseURL,
    apiVersion = 'v7',
    presets,
    ratio = 1.5,
    params = 'org_if_sml=1',
    imageSizeAttributes = 'use',
    exactSize = false,
    doNotReplaceURL = false,
    limitFactor = 100,
    devicePixelRatioList = CONSTANTS.DEVICE_PIXEL_RATIO_LIST,
    lowQualityPreview: {
      minImgWidth = 250,
    } = {},
  } = config;

  const [cloudImageConfig, setCloudImageConfig] = useState({
    token,
    domain,
    customDomain,
    lazyLoading,
    lazyLoadOffset,
    placeholderBackground,
    baseURL: baseUrl || baseURL,
    ratio,
    exactSize,
    presets: presets || {
      xs: '(max-width: 575px)', // to 575       PHONE
      sm: '(min-width: 576px)', // 576 - 767    PHABLET
      md: '(min-width: 768px)', // 768 - 991    TABLET
      lg: '(min-width: 992px)', // 992 - 1199   SMALL_LAPTOP_SCREEN
      xl: '(min-width: 1200px)', // from 1200    USUALSCREEN
    },
    params: processParams(params),
    apiVersion,
    innerWidth: typeof window !== 'undefined' ? window.innerWidth : null,
    previewQualityFactor: 10,
    doNotReplaceURL,
    devicePixelRatioList,
    limitFactor,
    minLowQualityWidth: minImgWidth,
    imageSizeAttributes,
  });

  const updateDimensions = debounce(100, () => {
    setCloudImageConfig({ ...cloudImageConfig, innerWidth: window.innerWidth });
  });

  useEffect(() => {
    if (typeof window !== 'undefined') {
      window.addEventListener('resize', updateDimensions);
    }

    return () => {
      window.removeEventListener('resize', updateDimensions);
    };
  }, []);

  return (
    <CloudimageContext.Provider
      value={useMemo(() => ({ cloudImageConfig }), [cloudImageConfig])}
    >
      {children}
    </CloudimageContext.Provider>
  );
}

export { CloudimageContext };
export default CloudimageProvider;
